src/hypothesis/extra/django/models.py

```diff
@@ -17,159 +17,402 @@
+    def __init__(self, model, db, model_data_strategy):
         super(ModelStrategy, self).__init__()
         self.model = model
-        self.arg_strategy = st.fixed_dictionaries(mappings)
+        self.db = db
+        self.model_data_strategy = model_data_strategy
 
     def __repr__(self):
         return u'ModelStrategy(%s)' % (self.model.__name__,)
 
     def do_draw(self, data):
         try:
-            result, _ = self.model.objects.get_or_create(
-                **self.arg_strategy.do_draw(data)
+            model_data = data.draw(self.model_data_strategy)
+        except UnmappedFieldError as ex:
+            raise InvalidArgument(
+                u'Missing argument for mandatory field {model}.{field}'
+                .format(
+                    model=self.model.__name__,
+                    field=ex.args[0].name,
+                )
             )
-            return result
-        except IntegrityError:
-            data.mark_invalid()
+        try:
+            # We need to wrap the model create in an atomic block, so
+            # we need to use the correct write database for the model.
+            db = data.draw(self.db) or router.db_for_write(self.model)
+            # If the save gives an IntegrityError, this will roll the
+            # savepoint back.
+            with transaction.atomic(using=db):
+                # Create the model inside the transaction, just in case it
+                # performs database actions in __init__.
+                obj = self.model(**model_data)
+                # This should catch unique checks, plus any other custom
+                # validation on the model.
+                obj.full_clean()
+                obj.save(using=db)
+            return obj
+        except DataError:
+            # Something weird happened. For example,
+            # some Postgres database encodings will refuse text that
+            # is longer than a VARCHAR *once encoded by the database*.
+            pass
+        except (ValidationError, IntegrityError) as ex:
+            # This might mean that a unique key was violated.
+            # As a fallback, try to get the identical model. This is needed
+            # for example() calls, which expect to be able to call the
+            # strategy idempotently.
+            model_keys = {
+                field_name: field_value
+                for field_name, field_value
+                in model_data.items()
+                # Remove auto fields from the model keys, since these
+                # will change with every instance.
+                if not isinstance(
+                    self.model._meta.get_field(field_name),
+                    dm.AutoField,
+                )
+            }
+            if model_keys:
+                try:
+                    return self.model._default_manager.get(**model_keys)
+                except self.model.DoesNotExist:
+                    # It was something other than a unique violation
```

Comment:
Would it make sense to handle `MultipleObjectsReturned` by choosing among the matching rows?